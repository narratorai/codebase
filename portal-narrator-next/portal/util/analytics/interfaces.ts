//////// PORTAL SPECIFIC PLUGIN TYPES: ////////
// - Override default ({ payload }) types in custom plugin definitions

export interface IdentifyPluginPayload {
  userId: string
  meta: any
  type: string
  traits: any
  // Portal specific options type:
  options: {
    user: any
    ldFlags: {
      [key: string]: boolean | string | number
    }
    companySlug?: string
  }
}

export interface TrackPluginPayload {
  anonymousId: string
  meta: any
  options: any
  event: string
  properties: any
}
