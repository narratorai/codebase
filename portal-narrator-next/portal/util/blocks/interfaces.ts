import { ObjectFieldTemplateProps, UiSchema } from '@rjsf/core'
import { JSONSchema7 } from 'json-schema'
import { INotification } from 'util/datasets/interfaces'
import { GetFileAPIReturn } from 'components/Narratives/interfaces'
import { ITableColumnOrder } from 'util/narratives/interfaces'
import { CONTENT_TYPE_METRIC_V2, CONTENT_TYPE_PLOT_V2, CONTENT_TYPE_TABLE_V2 } from 'util/narratives/constants'
import { ICompany_Table } from 'graph/generated'

//
// Service used to support getting schemas, submitting a form, etc
//

export interface LoadingBarProps {
  duration: number
  percent: number
  text: string
}

export interface RefreshConfig {
  field_slug?: string
  process_data?: boolean
  update_schema?: boolean
  submit_form?: boolean
  runImmediately?: boolean
  loading_bar?: LoadingBarProps[]
}

export interface IFormContext {
  onRefreshConfigRequest?: (refreshConfig: RefreshConfig) => void
  subscribeToRefresh?: ({
    fieldSlug,
    graphQuery,
    variables,
  }: {
    fieldSlug: string
    graphQuery: string
    variables: any
  }) => void
  version: number
}

export interface BlockService {
  loadSchemas({ asAdmin }: { asAdmin?: boolean }): Promise<IBlockOptions | undefined>
  submitForm(schemaSlug: string, data: DataSubmitBody, asAdmin?: boolean): Promise<BlockContent[]>
  // schema v1 endponts:
  loadBlock(body?: UpdateSchemaSubmitBody): Promise<UpdateSchemaResponse>
  updateSchema(schemaSlug: string, body?: UpdateSchemaSubmitBody, asAdmin?: boolean): Promise<UpdateSchemaResponse>
  processData(schemaSlug: string, body: ProcessDataSubmitBody, asAdmin?: boolean): Promise<ProcessDataResponse>
  loadItemContextById(id: string, resourceType: ResourceType, asAdmin?: boolean): Promise<UpdateSchemaResponse>
  loadDropdown(schemaSlug: string, body: ProcessDataSubmitBody, asAdmin?: boolean): Promise<LoadDropdownResponse>
}

// for loadItemContextById mavis api paths ex: v1/${resourceType}:
export type ResourceType = 'transformation' | 'activities'

export interface IBlockState {
  id: string
  resourceType: ResourceType
}

export interface LoadingBarOption {
  percent: number
  text: string
  duration: number
}

// What we get back as a list of potential blocks to select:
export interface GenericBlockOption {
  slug: string
  title: string
  version: number
  description?: string
  advanced: boolean
}

export type IBlockOptions = {
  [key in
    | 'blocks'
    | 'field_blocks'
    | 'narrative_blocks'
    | 'production_blocks'
    | 'field_loading_screen'
    | 'empty_dashboard'
    | 'empty_narrative']?: GenericBlockOption[] | LoadingBarOption[] | GetFileAPIReturn | null
}

// todo: this should be better specified
export type FormData = any

export interface FormSchema {
  schema: JSONSchema7
  ui_schema?: UiSchema
}

// Similar to IChangeEvent but formData --> data
// https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-jsonschema-form/index.d.ts
export interface FormState extends FormSchema {
  data: FormData
  internal_cache?: { [key: string]: any } // owned by the backend - we send it back to give it some context
}

export interface SubmitResult {
  content: BlockContent[]
  formData: FormData
}

export interface HandleRefreshStateArgs {
  field_slug: string
  update_schema: boolean
  process_data: boolean
  submit_form: boolean
  formData: FormData
  disableLoading?: boolean
  loading_bar?: LoadingBarProps[]
}

export interface UpdateSchemaSubmitBody {
  // technically we don't need any of these (including an empty body will intiialize an empty form)
  data?: FormData
  field_slug?: string
  block_slug?: string
  internal_cache?: { [key: string]: any }
  fields?: Record<string, unknown>[]
}

export type UpdateSchemaResponse = FormState

export interface CallBackResponse {
  type: string
  value: BlockContent[]
}

export interface ProcessDataSubmitBody {
  field_slug: string
  data: FormData
}

export interface ProcessDataResponse {
  data: FormData
  redirect_url?: string
  notification?: INotification
  dirty?: boolean
  show_beacon_id?: string
  confetti?: boolean
}

export type ProcessedResult = Omit<ProcessDataResponse, 'data'>

export type FieldValue = number | string
export type SelectOption = { value: FieldValue; label?: string }

export interface LoadDropdownResponse {
  data: {
    values: SelectOption[]
  }
}

export interface DataSubmitBody {
  data: FormData
}

export interface MarkdownContent {
  type: 'markdown'
  value: string
  text?: string
}

export interface TextContent {
  type: 'text'
  value: string
}

export interface JsonContent {
  type: 'json'
  value: {
    [key: string]: any
  }
}

export interface PlotContent {
  id: string
  type: 'plot' | 'block_plot' | typeof CONTENT_TYPE_PLOT_V2
  value: {
    data: { [key: string]: any }[]
    layout: {
      [key: string]: any
    }
    config: {
      [key: string]: any
    }
    height?: number
  }
}

export interface INarrativeTableColumn {
  name: string
  friendly_name: string
  format: string
  pinned: 'left' | 'right' | null
  type: string
}

export interface INarrativeTableContentMetaData {
  title?: string
  url?: string
  customer_column?: string
  customer_kind?: string
  retrieved_at?: string
  table?: Partial<ICompany_Table>
}

export interface INarrativeTableContent {
  columns: INarrativeTableColumn[]
  rows: Record<string, any>[]
  metadata?: INarrativeTableContentMetaData
}

export interface TableContent {
  type: 'table' | typeof CONTENT_TYPE_TABLE_V2
  value: INarrativeTableContent
  column_order?: ITableColumnOrder
}

export interface ImpactCalculatorContent {
  type: 'impact_calculator' | 'impact_simulator'
  value: { [key: string]: any }
}

export interface RawMectricContent {
  type: 'raw_metric' | typeof CONTENT_TYPE_METRIC_V2
  value: { [key: string]: any }
}

export interface AnalyzeSimulatorContent {
  type: 'analyze_simulator'
  value: {
    default_shift_percent: number
    previous_names: string[]
    requester: {
      email: string
    }
    kpi_name: string
    current_kpi: number
    higher_is_good: boolean
    row_label: string
    shift_from_group: string
    shift_from_kpi: number
    shift_from_percent: number
    shift_to_group: string
    shift_to_percent: number
    shift_to_kpi: number
    _raw_fields: {}
    simulations: {
      percent_shift: number
      simulated_kpi: number
      lift: number
    }[]
  }
}

export type BlockContent =
  | MarkdownContent
  | JsonContent
  | TextContent
  | PlotContent
  | TableContent
  | ImpactCalculatorContent
  | RawMectricContent
  | AnalyzeSimulatorContent

export type BlockContentType =
  | 'markdown'
  | 'text'
  | 'json'
  | 'plot'
  | 'block_plot'
  | 'table'
  | 'impact_calculator'
  | 'raw_metric'
  | 'analyze_simulator'

//
// Support for ui:tabs ObjectFieldTemplate integration
//

export interface TabConfig {
  label: string
  property_names: string[]
  tab_id: string
  redirect_tab_ids: string[]
  'ui:info_modal'?: string
}

export interface TabsConfig {
  other_label?: string
  tabs: TabConfig[]
}

export interface GroupedTab {
  label: string
  properties: ObjectFieldTemplateProps['properties']
  tab_id: string
  redirect_tab_ids: string[]
}

export interface FieldConfig {
  name: string
  kind: string
  value: any
  format?: any
  explanation?: string
  unused?: boolean
  field_depends_on?: string[]
  previous_names?: string[]
}
