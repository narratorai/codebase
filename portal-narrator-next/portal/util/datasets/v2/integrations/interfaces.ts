export interface IIntegrationsConfig {
  [key: string]: {
    displayName: string
    description?: string
    disableParentGroup?: boolean
  }
}

export type BiToolType = 'looker' | 'tableau' | 'metabase' | 'powerbi' | 'datastudio' | 'domo' | 'otherbi'
