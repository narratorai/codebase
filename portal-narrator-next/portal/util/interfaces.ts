// TODO switch uses of these to graph/generated types!

export type GetToken = () => Promise<string | undefined>
export interface IGraphActivity {
  id: string
  name: string
  slug: string
}

export interface IActivity {
  _id: string
  _last_modified_at: string
  name: string
  slug: string | string[]
  table: string
  script_slugs: string[]
  // TODO - make these not any:
  features: any[]
  meta: any
}

// the MAVIS owned company config file that lives in a company's S3 bucket:
export interface ICompanyConfig {
  production_schema: string
  tables: IConfigStreamTable[]
  timezone: string
  spend_table: string
}

// An instance of a stream table and it's meta fields:
export interface IConfigStreamTable {
  activity_stream: string
  customer_label: string
  customer_table: string
  identifier: string
}

export interface INarrative {
  id: string
  name: string
  slug: string
  company: string
  created_at: string
  description: string
  state: 'in_progress' | 'live' | 'on_stream_update' | 'cron'
  template_global_slug: string
  update_cron: string
  update_kind: string
}
