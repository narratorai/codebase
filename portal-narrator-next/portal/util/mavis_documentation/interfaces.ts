export type MavisDocumentationResponse = {
  all_documents?: [
    {
      markdown: string
    }
  ]
}
