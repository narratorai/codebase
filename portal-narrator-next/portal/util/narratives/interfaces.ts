import { IContent } from 'components/Narratives/interfaces'
import type narrativeFromTemplateMachine from 'machines/narrativeTemplates/narrativeFromTemplateMachine'
import { FieldConfig } from 'util/blocks/interfaces'
import { IDatasetQueryDefinition } from 'util/datasets/interfaces'
import { CONTENT_TYPE_IMAGE_UPLOAD, CONTENT_TYPE_MEDIA_UPLOAD } from 'util/narratives/constants'
import { StateFrom } from 'xstate'

export type BlockType =
  | 'markdown'
  | 'metric_v2'
  | 'plot_v2'
  | 'table_v2'
  | 'simple_plot'
  | 'narrative_plotter'
  | 'impact_calculator'
  | 'simple_metric'
  | 'raw_metric'
  | 'dataset_table'
  | 'raw_table'
  | 'csv_table'
  | 'analyze_simulator'
  | typeof CONTENT_TYPE_IMAGE_UPLOAD
  | typeof CONTENT_TYPE_MEDIA_UPLOAD

export interface BaseMapping {
  new_id: string | null
  old_id: string
  question: string
}

interface FeatureMapping extends BaseMapping {
  display_name: string | null
  allowed_types: string[]
}

export interface DatasetFeatureMapping extends FeatureMapping {
  _dataset_mapping_old_id: string
}

export interface TemplateDatasetConfig {
  name: string
  dataset: IDatasetQueryDefinition
  feature_mapping: FeatureMapping[]
  mapping: BaseMapping
}

// FIXME
// FIXME - is it activity_mappings or activity_mapping???????
// FIXME - is it word_mappings or word_mapping???????
// FIXME - is it feature_mappings or feature_mapping???????
// FIXME

export interface BaseNarrativeTemplateConfig {
  activity_mapping: BaseMapping[]
  datasets: TemplateDatasetConfig[]
  word_mappings: BaseMapping[]
  additional_context?: string
}

export interface INarrativeTemplateConfig extends BaseNarrativeTemplateConfig {
  // TODO: add types to the following:
  narrative: {
    datasets: any[]
    field_configs: FieldConfig[]
    narrative: INarrativeConfig
    _last_modified_at: string
  }
}

//////////////////////// Mavis API Response Types ////////////////////////

export interface FeatureOption {
  label: string
  value: string
}

export interface FeatureMappingOption {
  feature_mapping_old_id: string
  options: FeatureOption[]
}

export interface FeatureOptionsData {
  feature_mapping_options: FeatureMappingOption[]
}

export interface CreateNarrativeResponse {
  narrative_slug: string
  dataset_slugs: string[]
}
export interface ITemplateContext {
  machineCurrent: StateFrom<typeof narrativeFromTemplateMachine>
  machineSend: Function
}

export type NarrativeFields = Record<string, string | number | null>

export interface INarrativeConfig {
  question?: string
  goal?: string
  recommendation?: {
    title?: string
    explanation?: string
  }
  is_actionable?: boolean | string | number | null
  key_takeaways?: {
    explanation?: string
    title?: string
    value?: number
  }[]
  sections?: {
    id: string
    content?: IContent[]
    show?: boolean
    takeaway?: string | null
    title?: string | null
    _dashboard_layout_version?: number
  }[]
}

export interface IAssembledFieldsResponse {
  updated?: string[]
  refresh?: boolean
  fields?: NarrativeFields
  field_configs?: FieldConfig[]
  narrative?: INarrativeConfig
}

export interface IDependencyGraphResponse {
  branches: {
    id: string
    source: string
    target: string
  }[]
  nodes: {
    id: string
    data: {
      failed: boolean
      label: string
      kind: string
      node_kind: string
      unused: boolean
      value: string
      x_position: number
      y_position: number
    }
  }[]
}

export interface ITableColumnOrder {
  left: string[]
  right: string[]
  order: string[]
}
