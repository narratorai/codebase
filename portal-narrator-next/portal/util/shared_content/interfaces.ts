import {
  COPIED_PLOT_CONTENT_TYPE,
  COPIED_METRIC_CONTENT_TYPE,
  COPIED_TABLE_CONTENT_TYPE,
} from 'util/shared_content/constants'

export interface CopiedPlotContent {
  id: string
  type: typeof COPIED_PLOT_CONTENT_TYPE
  data: {
    dataset_slug: string
    group_slug: string
    plot_slug: string
  }
}

export interface CopiedMetricContent {
  id: string
  type: typeof COPIED_METRIC_CONTENT_TYPE
  data: {
    dataset_slug: string
    group_slug: string
    column_id: string
  }
}

export interface CopiedTableContent {
  id: string
  type: typeof COPIED_TABLE_CONTENT_TYPE
  data: {
    dataset_slug: string
    group_slug: string
    as_data_table: boolean
    limit?: number | null
    title?: string | null
  }
}
