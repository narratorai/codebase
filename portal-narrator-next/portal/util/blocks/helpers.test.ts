import React from 'react'
import {
  findBlockContentByType,
  findAllBlockContentByType,
  groupPropertiesByTab,
  getAllDependentFields,
  getDependentFieldsForField,
} from './helpers'
import { BlockContent, PlotContent, MarkdownContent, JsonContent } from './interfaces'
import fieldConfigs from '../../../test/fixtures/fieldConfigs.json'

describe('util/blocks/helpers', () => {
  describe('#findBlockContentByType', () => {
    it('handles `null` as blockContent', () => {
      expect(findBlockContentByType<any>('plot', null)).toEqual(undefined)
    })

    it('returns plot content based on type', () => {
      const blockContent: BlockContent[] = [
        {
          type: 'plot',
          value: {
            data: [{}],
            layout: {},
            config: {},
          },
        },
        {
          type: 'markdown',
          value: '#hello',
        },
      ]

      const { type, value } = findBlockContentByType<PlotContent>('plot', blockContent) || {}
      const data = value?.data

      expect(type).toEqual('plot')
      expect(data).toEqual([{}])
    })

    it('returns markdown content based on type', () => {
      const blockContent: BlockContent[] = [
        {
          type: 'plot',
          value: {
            data: [{}],
            layout: {},
            config: {},
          },
        },
        {
          type: 'markdown',
          value: '#hello',
        },
      ]

      const { type, value } = findBlockContentByType('markdown', blockContent) as MarkdownContent

      expect(type).toEqual('markdown')
      expect(value).toEqual('#hello')
    })
  })

  describe('#findAllBlockContentByType', () => {
    it('handles `null` as blockContent', () => {
      expect(findAllBlockContentByType<any>('plot', null)).toEqual([])
    })

    it('returns plot content based on type', () => {
      const blockContent: BlockContent[] = [
        {
          type: 'plot',
          value: {
            data: [{}],
            layout: {},
            config: {},
          },
        },
        {
          type: 'markdown',
          value: '#hello',
        },
      ]
      const resp = findAllBlockContentByType<PlotContent[]>('plot', blockContent)
      const type = resp?.[0]?.type
      const value = resp?.[0]?.value
      const data = value?.data

      expect(type).toEqual('plot')
      expect(data).toEqual([{}])
    })

    it('returns multiple plot content based on type', () => {
      const blockContent: BlockContent[] = [
        {
          type: 'plot',
          value: {
            data: [{}],
            layout: {},
            config: {},
          },
        },
        {
          type: 'markdown',
          value: '#hello',
        },
        {
          type: 'plot',
          value: {
            data: [{}],
            layout: {},
            config: {},
          },
        },
      ]

      const resp = findAllBlockContentByType<PlotContent[]>('plot', blockContent)

      const type1 = resp?.[0]?.type
      const value1 = resp?.[0]?.value
      const type2 = resp?.[1]?.type
      const value2 = resp?.[1]?.value

      const data1 = value1?.data
      const data2 = value2?.data

      expect(type1).toEqual('plot')
      expect(data1).toEqual([{}])
      expect(type2).toEqual('plot')
      expect(data2).toEqual([{}])
      expect(resp?.length).toEqual(2)
    })

    it('returns multiple markdown content based on type', () => {
      const blockContent: BlockContent[] = [
        {
          type: 'markdown',
          value: '#hello',
        },
        {
          type: 'plot',
          value: {
            data: [{}],
            layout: {},
            config: {},
          },
        },
        {
          type: 'markdown',
          value: '#hello',
        },
      ]

      const resp = findAllBlockContentByType<MarkdownContent[]>('markdown', blockContent)

      const type1 = resp?.[0]?.type
      const value1 = resp?.[0]?.value
      const type2 = resp?.[1]?.type
      const value2 = resp?.[1]?.value

      expect(type1).toEqual('markdown')
      expect(value1).toEqual('#hello')
      expect(type2).toEqual('markdown')
      expect(value2).toEqual('#hello')
      expect(resp?.length).toEqual(2)
    })

    it('returns multiple json content based on type', () => {
      const blockContent: BlockContent[] = [
        {
          type: 'json',
          value: {
            name: 'first_record_count_records',
            format: {},
            kind: 'dataset_metric',
            value: {
              dataset_slug: 'testing_state_machine',
              group_slug: 'group_2',
            },
            field_depends_on: [],
          },
        },
        {
          type: 'markdown',
          value: '#hello',
        },
        {
          type: 'json',
          value: {
            name: 'first_record_count_records',
            format: {},
            kind: 'dataset_metric',
            value: {
              dataset_slug: 'testing_state_machine',
              group_slug: 'group_2',
            },
            field_depends_on: [],
          },
        },
      ]

      const resp = findAllBlockContentByType<JsonContent[]>('json', blockContent)

      const type1 = resp?.[0]?.type
      const value1 = resp?.[0]?.value
      const type2 = resp?.[1]?.type
      const value2 = resp?.[1]?.value

      const name1 = value1?.name
      const name2 = value2?.name

      expect(type1).toEqual('json')
      expect(name1).toEqual('first_record_count_records')
      expect(type2).toEqual('json')
      expect(name2).toEqual('first_record_count_records')
      expect(resp?.length).toEqual(2)
    })
  })

  describe('#groupPropertiesByTab', () => {
    it('returns grouped properties by tab', () => {
      const tabsConfig = {
        tabs: [
          {
            label: 'Tab A',
            property_names: ['slug', 'notes'],
            tab_id: 'tab_a',
            redirect_tab_ids: [],
          },
        ],
      }
      const properties = [
        { content: React.createElement('div'), name: 'slug', disabled: false, readonly: false },
        { content: React.createElement('div'), name: 'other_a', disabled: false, readonly: false },
        { content: React.createElement('div'), name: 'notes', disabled: false, readonly: false },
        { content: React.createElement('div'), name: 'other_b', disabled: false, readonly: false },
        { content: React.createElement('div'), name: 'other_c', disabled: false, readonly: false },
      ]
      expect(groupPropertiesByTab(tabsConfig, properties)).toMatchSnapshot()
    })
  })

  describe('#getAllDependentFields', () => {
    it('should return all dependent fields for a set of fields', () => {
      const fields = [
        {
          name: 'engagement_rate_local_trvl',
          kind: 'value',
          format: 'percent',
          value: '$$eval({engagement_rate}*(1+{percent_diff_in_distribution_local_trvl}))',
          field_depends_on: ['engagement_rate', 'percent_diff_in_distribution_local_trvl'],
        },
        {
          name: 'dist_all_category',
          kind: 'value',
          format: 'percent',
          value: '$$eval(1-{percent_of_LCL_trvl})',
          field_depends_on: ['percent_of_LCL_trvl'],
        },
      ]

      const expected = ['engagement_rate', 'percent_diff_in_distribution_local_trvl', 'percent_of_LCL_trvl']
      expect(getAllDependentFields(fields)).toEqual(expect.arrayContaining(expected))
    })

    it('should return an empty array for fields with no field_depends_on values', () => {
      const fields = [
        {
          name: 'TRVL_percent_of_total_taps',
          kind: 'dataset_metric',
          format: 'percent',
          value: {
            dataset_slug: 'content_viewed_before_transactions',
          },
          field_depends_on: [],
        },
        {
          name: 'TRVL_percent_of_total_taps_w_transactions',
          kind: 'dataset_metric',
          format: 'percent',
          value: {
            dataset_slug: 'content_viewed_before_transactions',
          },
          field_depends_on: [],
        },
      ]

      expect(getAllDependentFields(fields)).toEqual([])
    })
    it('should return an empty array when passed an empty array', () => {
      expect(getAllDependentFields([])).toEqual([])
    })
    it('should match snapshot', () => {
      expect(getAllDependentFields(fieldConfigs)).toMatchSnapshot()
    })
  })

  describe('#getDependentFieldsForField', () => {
    it('should return an empty array for a field that is NOT referenced by any other field', () => {
      const fieldConfigs = [
        {
          name: 'last_month_txns',
          kind: 'dataset_metric',
          format: {},
          value: {
            dataset_slug: 'content_and_transactions',
          },
          field_depends_on: ['last_month'],
        },
        {
          name: 'last_month_conv_to_completed_transaction',
          kind: 'dataset_metric',
          format: 'percent',
          value: {
            dataset_slug: 'content_and_transactions',
          },
          field_depends_on: ['last_month'],
        },
      ]

      expect(getDependentFieldsForField({ fieldConfigs, fieldConfig: fieldConfigs[0] })).toEqual([])
    })
    it('should return an array of fields that reference a field', () => {
      const fieldConfigs = [
        {
          name: 'percent_of_LCL_trvl',
          kind: 'dataset_metric',
          format: 'percent',
          value: {
            dataset_slug: 'card_taps_with_swipes',
          },
          field_depends_on: [],
        },
        {
          name: 'dist_all_category',
          kind: 'value',
          format: 'percent',
          value: '$$eval(1-{percent_of_LCL_trvl})',
          field_depends_on: ['percent_of_LCL_trvl'],
        },
      ]

      expect(getDependentFieldsForField({ fieldConfigs, fieldConfig: fieldConfigs[0] })).toEqual(['dist_all_category'])
    })

    it('should match the snapshot', () => {
      // fieldConfigs[22] has name: 'last_month_total_taps'
      expect(getDependentFieldsForField({ fieldConfigs, fieldConfig: fieldConfigs[22] })).toMatchSnapshot()
    })
  })
})
