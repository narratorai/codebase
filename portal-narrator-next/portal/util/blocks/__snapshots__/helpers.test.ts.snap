// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/blocks/helpers #getAllDependentFields should match snapshot 1`] = `
[
  "last_month",
  "max_conversion_rate_to_completed_transaction",
  "min_conversion_rate_to_completed_transaction",
  "max_category_tapped",
  "last_month_total_swipes",
  "last_month_total_taps",
  "engagement_rate",
  "percent_diff_in_distribution_local_trvl",
  "percent_of_LCL_trvl",
]
`;

exports[`util/blocks/helpers #getDependentFieldsForField should match the snapshot 1`] = `
[
  "cards_shown",
  "engagement_rate",
]
`;

exports[`util/blocks/helpers #groupPropertiesByTab returns grouped properties by tab 1`] = `
[
  {
    "label": "Tab A",
    "properties": [
      {
        "content": <div />,
        "disabled": false,
        "name": "slug",
        "readonly": false,
      },
      {
        "content": <div />,
        "disabled": false,
        "name": "notes",
        "readonly": false,
      },
    ],
    "property_names": [
      "slug",
      "notes",
    ],
    "redirect_tab_ids": [],
    "tab_id": "tab_a",
  },
]
`;
